### PR TITLE
Added better .wcon file examples 

### PR DESCRIPTION
**Pull request by MichaelCurrie**

### CODE REVIEW

**Comment by MichaelCurrie** [comment]
This handles #6.  @Ichoran please merge if you agree, or let me know if you think more work needs to be done.

For `maximal_1.wcon`, etc, I created three chunks rather than two so we could test the case when neither of `"prev"` or `"next"` is `null`.  Also this example is right now not really "maximal" in the sense of having lots of complicated entries in the "data" object so it could be added to in the future.

File: tests/minimal.wcon [hunk 1]
```diff
@@ -0,0 +1 @@
+{"tracker-commons":true, "units":{}, "data":[]}
```
Comment (Ichoran) [inline tests/minimal.wcon]: I think we require units on "x", "y",  and "t".  That we don't use them in the data doesn't enter the specification.

File: tests/maximal_0.wcon [hunk 2]
```diff
@@ -0,0 +1 @@
+{"tracker-commons":true,"files":{"this":"_0", "prev":null, "next":["_1", "_2"]},"units":{},"data":[{"id":3, "t":1.3, "x":[3,4], "y":[5.4,3]}]}
```
Comment (Ichoran) [inline tests/maximal_0.wcon]: I don't think units are optional in the spec.  This should fail on a strict reader.

File: tests/minimax.wcon [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+           "protocol":"text description of protocol",
+           "software":{
+                "tracker":{ "name":"Software Name", "version":"1.3.0"},
+                "featureID":"@OMG"
+           },
+           "settings":"Any valid JSON entry with hardware and software configuration can go here"
+     },
+    "units":{"t":"s", "x":"mm", "y":"m"},
+    "data":[
+        {"@OMG":{
```
Comment (Ichoran) [inline tests/minimax.wcon]: Did the spec get rewritten to allow entries that don't have "id" etc.?  Because JSON does not require Objects to have keys in any particular order, there is no way to tell that this entry doesn't have "id" etc. without reading the whole thing.  What is the rule supposed to be?  Personally, I would put the whole `@OMG` section outside of "data" (which is okay by spec).

File: tests/minimax.wcon [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        }},	
+        { "id":4, "t":1.4, "x":[1215.11, 1216.14, 1217.12], "y":[234.89, 265.23, 235.08] },
+	    { "id":1, "t":[1.3,1.5], "x":[[1,1,1],[1215.11, 1216.14, 1217.12]], 
+	                                 "y":[[2,2,2],[234.89, 265.23, 235.08]], "ox":[5000,5001] },
+		{ "id":2, "t":1.4, "x":[125.11, 126.14, 117.12], "y":[23.3, 22.23, 21135.08] },
+		{ "id":1, "t":1.4, "x":[1215.11, 1216.14, 1217.12], "y":[234.89, 265.23, 235.08]},
+		{ "id":1, "t":1.3, "x":[1, 1, 1], "y":[2, 2, 2], "@OMG":{"feature_values":[55,55,55]}},
+		{ "id":2, "t":1.5, "x":[1215.11, 1216.14, 1217.12], "y":[234.89, 265.23, 235.08] }
+	],
+	"@AQ": 5
```
Comment (Ichoran) [inline tests/minimax.wcon]: That is, why isn't `@OMG` at this level too?  (`@AQ` is totally fine here.)

**Comment by Ichoran** [comment]
I agree that we need more than two so we get all the variants on prev/next.  This looks good save for the in-line comments I left.

File: tests/minimal.wcon (hunk 1 above)
Comment (MichaelCurrie) [inline tests/minimal.wcon]: I read the specification as being silent on which properties are required, only that the `"units"` object itself is required.

So, I've amended the `WCON_format.md` to be more explicit:

From

> `"units"` is required and must be single-valued.

to

> `"units"` is required and must be single-valued.  Units must be specified at least for `"t"`, `"x"`, and `"y"`.

I've also changed all my tests and examples to reflect this.

File: tests/maximal_0.wcon (hunk 2 above)
Comment (MichaelCurrie) [inline tests/maximal_0.wcon]: OK, I've changed this to add x y and t properties here for this units object.

File: tests/minimax.wcon (hunk 3 above)
Comment (MichaelCurrie) [inline tests/minimax.wcon]: The spec does say that "id" is required:

> A data entry is an object with four required keys.

However, later, in the section on custom features, there seem to be three places they can be located.  They must all start with "@".
1. As a top-level object with a key starting with "@"
2. As an element of the "data" array.  This is the one you are concerned about, I believe.
3. As a property of an element of the "data" array, alongside "id", "t", "x", "y".

You can see all three examples in this code excerpt from the spec:

> ``` JSON
> {
>     "tracker-commons":true,
>     "units":{"t":"s", "x":"mm", "y":"mm"},
>     "data":[
>         {"@OMG":{
>                "feature_names":["speed", "curvature", "width"],
>                "feature_units":["mm/s", "1/mm", "mm"]
>         }},
>         { "id":1,
>           "t":1.3,
>           "x":[1215.11, ...],
>           "y":[234.89, ...],
>           "@OMG":{ "feature_values":[0.34, 1.5, 0.873] }
>         }
>         ...
>     ],
>     "@OMG":{
>           "plate_feature_names":["density", "aggregate number"],
>           "plate_feature_units":["1/mm^2", null],
>           "plate_feature_values":[0.1, 8]
>     }
> }
> ```

I would be in favour of simplifying the spec to include just type 1, the top-level custom feature, because in that top-level custom feature other labs could easily construct a structure parallel to the "data" object, giving all the same time-series data, or any other kind of data they want.

Please let me know if you agree and I'll modify the spec document and my code.

File: tests/minimax.wcon (hunk 4 above)
Comment (MichaelCurrie) [inline tests/minimax.wcon]: I think your comment here might be moot if you accept my previous comment's recommendations to use only top-level custom features.

File: tests/minimax.wcon (hunk 3 above)
Comment (aexbrown) [inline tests/minimax.wcon]: I would prefer to keep as much flexibility as possible with the custom features.  One case where it makes sense to include a feature within data is if you do 3D tracking and you have a z-coordinate.  It feels unnatural to have one of three coordinates in a separate object.  On the other hand, almost no one is doing 3D tracking right now.  If that's really the only case we can come up with where it's a problem, then maybe we should go for the simplified spec.

Opinions?

Comment (Ichoran) [inline tests/minimax.wcon]: The problem with `[{"@OMG": {}, {"id"=1, "t"=1, "x"=2, "y"=3 }]` is that the data array is no longer an array of comprehensible items.  The problem dropping `[{"id"=1, "t"=1, "x"=2, "y"=3, "@XJ": {"outline": "0zxqjhFLhqgf20%1fjqFDq75"}}]` is that some people (e.g me) need to store extra data on every timepoint for every animal.

I suppose we could say that the ID is optional, so you can put things in the data array that aren't animals (and only things with an ID are required to have x and y).  But if you don't even have a time, I'm not sure why it belongs in the data array.  I like flexibility, but I want the format to be constrained enough so you at least have some idea of where to look for which kind of data.

File: src/Python/tests.py [hunk 5]
```diff
@@ -94,104 +94,104 @@ def test_tracker_commons_and_units(self):
...
 
         # The smallest valid WCON file (Empty data array should be fine)
-        WCONWorm.load(StringIO('{"tracker-commons":true, "units":{}, "data":[]}'))
+        WCONWorm.load(StringIO('{"tracker-commons":true, "units":{"t":"s","x":"mm","y":"mm"}, "data":[]}'))
 
         # Duplicate keys should cause the parser to fail
         with self.assertRaises(KeyError):
             WCONWorm.load(StringIO('{"tracker-commons":true, '
-                                    '"units":{"t":"s", "t":"s"}'
+                                    '"units":{"t":"s","t":"s","x":"mm","y":"mm"}'
```
Comment (Ichoran) [inline src/Python/tests.py:116]: This line has `"t":"s"` twice for some reason.

File: tests/minimax.wcon (hunk 3 above)
Comment (MichaelCurrie) [inline tests/minimax.wcon]: Having extra data on every timepoint for every animal, such as a "z", would indeed by nice to have within the "data" object itself.  So it sounds like we'd like to keep special feature type 1 and type 3, and drop type 2.

Note that the spec allows for units for more than just "x", "y", and "t", so 

```
"data":[
    {"@OMG":{
               "feature_names":["speed", "curvature", "width"],
               "feature_units":["mm/s", "1/mm", "mm"]
    }}
]
```

is unnecessary anyway.  We can just add those units, to the top-level "units" object:

```
"units": {
    "x":"mm",
    "y":"mm",
    "t":"s",
    "speed":"mm/s",
    "curvature":"1/mm",
    "width":"mm"
}
```

So, in summary, I agree that dropping the type 2 special feature would be better.  I've altered the English and JSON specification, test cases, and Python code to reflect this.

@Ichoran please feel free to merge this pull request if you are satisfied with the above.  Thanks.

File: src/Python/tests.py (hunk 5 above)
Comment (MichaelCurrie) [inline src/Python/tests.py:116]: This is intentional; I'm testing to verify that it will fail on that erroneous data.

(viz. `with self.assertRaises(KeyError):` on the previous line)

**Comment by MichaelCurrie** [comment]
I'll add my comment here since it seems to get hidden above:

Having extra data on every timepoint for every animal, such as a "z", would indeed by nice to have within the "data" object itself.  So it sounds like we'd like to keep special feature type 1 and type 3, and drop type 2.

Note that the spec allows for units for more than just "x", "y", and "t", so 

```
"data":[
    {"@OMG":{
               "feature_names":["speed", "curvature", "width"],
               "feature_units":["mm/s", "1/mm", "mm"]
    }}
]
```

is unnecessary anyway.  We can just add those units, to the top-level "units" object:

```
"units": {
    "x":"mm",
    "y":"mm",
    "t":"s",
    "speed":"mm/s",
    "curvature":"1/mm",
    "width":"mm"
}
```

So, in summary, I agree that dropping the type 2 special feature would be better.  I've altered the English and JSON specification, test cases, and Python code to reflect this.

@Ichoran please feel free to merge this pull request if you are satisfied with the above.  Thanks.
